File: dnn-providers/hip-kernel-provider/rocke/platform/python/rocke/helpers/mma_io.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
"""Descriptor-driven packed transport, expanded before IR serialization."""

from __future__ import annotations

from collections.abc import Callable
from math import gcd

from ..core.dtypes import dtype_info
from ..core.ir import (
    I8,
    I16,
    I32,
    I64,
    IRBuilder,
    Type,
    Value,
    VectorType,
    dtype_to_ir_type,
)
from ..core.storage import FragmentPacking, MatrixFragmentLayout, TensorStorage


def storage_ir_type(dtype: str) -> Type:
    """Addressable storage unit, distinct from the logical dtype resolver."""
    info = dtype_info(dtype)
    if info.encoded_bits % 8 or info.name in ("e8m0", "e4m3", "e5m3"):
        return I8
    return dtype_to_ir_type(info.name)


def load_matrix_fragment(
    b: IRBuilder,
    ptr: Value,
    row_base: Value,
    lane_group: Value,
    k0: int,
    *,
    storage: TensorStorage,
    layout: MatrixFragmentLayout,
    carrier_type: Type = I32,
) -> Value:
    """Load whole-byte chunks directly into the atom's carrier vector.

    row_base counts pointer storage units. lane_group is derived from the atom's
    lane mapping by the caller. Partial chunks and nonzero bit origins reject.
    """
    packing = layout.fragment
    if not (0 < packing.count <= 0x7FFFFFFF and packing.carrier_count <= 0x7FFFFFFF):
        raise ValueError("invalid matrix fragment chunk layout")
    unit_type = storage_ir_type(storage.dtype)
    unit_bytes = dtype_info(unit_type.name).encoded_bits // 8
    if ptr.type.pointee != unit_type:
        raise ValueError("matrix pointer storage type mismatch")
    if storage.packing != packing.packing or storage.base_bit_offset:
        raise ValueError(
            "matrix fragment requires matching packing and a zero bit origin"
        )
    span = layout.chunk_elements * layout.lane_groups * layout.chunks_per_lane
    if k0 < 0 or k0 + span > storage.shape[1]:
        raise ValueError("matrix fragment exceeds the packed row")
    origin_bits = storage.packing.bit_offset(k0)
    if origin_bits % (8 * unit_bytes) or layout.chunk_bytes % unit_bytes:
        raise ValueError("matrix fragment is not aligned to pointer storage units")
    if dtype_info(carrier_type.name).encoded_bits != packing.carrier_bits:
        raise ValueError("matrix carrier type width mismatch")
    if packing.payload_bits % packing.carrier_bits:
        raise ValueError("matrix fragment payload must occupy whole carriers")
    chunk_units = layout.chunk_bytes // unit_bytes
    origin_bytes = origin_bits // 8
    if (
        origin_bytes // unit_bytes > 0x7FFFFFFF
        or chunk_units * layout.lane_groups > 0x7FFFFFFF
    ):
        raise ValueError("matrix fragment offset exceeds i32 range")
    alignment = gcd(
        storage.alignment_bytes,
        storage.row_stride_bytes,
        layout.chunk_bytes,
        origin_bytes,
    )
    lane_chunk = b.mul(lane_group, b.const_i32(chunk_units))
    step_base = b.add(row_base, b.const_i32(origin_bytes // unit_bytes))
    chunks = []
    for j in range(layout.chunks_per_lane):
        offset = b.add(
            b.add(step_base, b.const_i32(j * layout.lane_groups * chunk_units)),
            lane_chunk,
        )
        remaining = chunk_units
        consumed = 0
        max_width = 8 if unit_bytes == 4 else 16
        while remaining:
            width = min(max_width, 1 << (remaining.bit_length() - 1))
            at = b.add(offset, b.const_i32(consumed)) if consumed else offset
            load_align = gcd(alignment, consumed * unit_bytes)
            if width == 1:
                value = b.vector_splat(
                    b.global_load(ptr, at, unit_type, align=load_align), 1
                )
            else:
                value = b.global_load_vN(ptr, at, unit_type, width, align=load_align)
            chunks.append(value)
            consumed += width
            remaining -= width
    payload = chunks[0]
    for chunk in chunks[1:]:
        payload = b.vec_concat(payload, chunk)
    payload_type = VectorType(carrier_type, packing.live_carriers)
    if payload.type != payload_type:
        payload = b.bitcast(payload, payload_type)
    padding = packing.carrier_count - packing.live_carriers
    if padding:
        if carrier_type != I32:
            raise ValueError("padded matrix fragments currently require i32 carriers")
        payload = b.vec_concat(payload, b.vector_splat(b.const_i32(0), padding))
    return payload


def pack_fragment_bits(
    b: IRBuilder,
    load_bits: Callable[[int], Value],
    fragment: FragmentPacking,
) -> list[Value]:
    """Pack unsigned encoded patterns into integer carriers, including split fields.

    The loader must return canonical patterns with zero high bits. Tensor decoding
    and numeric quantization are separate operations.
    """
    if fragment.carrier_bits not in (32, 64):
        raise ValueError("IR pattern packing currently requires i32 or i64 carriers")
    word_type = I64 if fragment.carrier_bits == 64 else I32
    constant = b.const_i64 if fragment.carrier_bits == 64 else b.const_i32
    words = [constant(0) for _ in range(fragment.carrier_count)]
    for j in range(fragment.count):
        pattern = load_bits(j)
        if pattern.type not in (I8, I16, I32, I64):
            raise ValueError(
                "pattern packing requires unsigned patterns in integer carriers"
            )
        pattern_bits = dtype_info(pattern.type.name).encoded_bits
        if pattern_bits < fragment.packing.element_bits:
            raise ValueError("pattern carrier is smaller than encoded width")
        if pattern_bits < fragment.carrier_bits:
            pattern = b.zext(pattern, word_type)
        elif pattern_bits > fragment.carrier_bits:
            raise ValueError("pattern carrier is wider than output carrier")
        start = fragment.packing.bit_offset(j)
        remaining = fragment.packing.element_bits
        consumed = 0
        while remaining:
            word, shift = divmod(start, fragment.carrier_bits)
            take = min(remaining, fragment.carrier_bits - shift)
            part = b.lshr(pattern, constant(consumed)) if consumed else pattern
            if take < remaining:
                part = b.land(part, constant((1 << take) - 1))
            words[word] = b.lor(words[word], b.shl(part, constant(shift)))
            start += take
            consumed += take
            remaining -= take
    return words

File: dnn-providers/hip-kernel-provider/rocke/platform/cpp/helpers/mma_io.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/* Descriptor-driven transport. Mirrors helpers/mma_io.py in emission order. */
#include "rocke/helper_rocke.helpers.mma_io.h"

#include <limits.h>
#include <string.h>

#include "rocke/error_boundary.hpp"

const rocke_type_t* rocke_storage_ir_type(const char* dtype)
{
    const rocke_dtype_info_t* info = rocke_dtype_info(dtype);
    if(!info)
        return NULL;
    if(info->encoded_bits % 8 || strcmp(info->name, "e8m0") == 0 || strcmp(info->name, "e4m3") == 0
       || strcmp(info->name, "e5m3") == 0)
        return rocke_i8();
    return rocke_dtype_to_ir_type(info->name);
}

static uint64_t gcd(uint64_t a, uint64_t b)
{
    while(b)
    {
        uint64_t next = a % b;
        a = b;
        b = next;
    }
    return a;
}

rocke_value_t* rocke_h_load_matrix_fragment(rocke_ir_builder_t* b,
                                            rocke_value_t* ptr,
                                            rocke_value_t* row_base,
                                            rocke_value_t* lane_group,
                                            uint64_t k0,
                                            const rocke_tensor_storage_t* storage,
                                            const rocke_matrix_fragment_layout_t* layout,
                                            const rocke_type_t* carrier_type)
{
    return ckc::guard_builder(b, [&]() -> rocke_value_t* {
        if(!rocke_i_live(b))
            return NULL;
        if(!storage || !storage->dtype || !layout || !carrier_type || !ptr || !row_base
           || !lane_group)
            ckc::raise_status(ROCKE_ERR_VALUE, "null matrix fragment argument");
        const auto& packing = layout->fragment;
        rocke_matrix_fragment_layout_t checked;
        if(!rocke_matrix_fragment_layout_init(&checked,
                                              &packing,
                                              layout->chunk_elements,
                                              layout->lane_groups,
                                              layout->lanes_per_group)
           || !packing.count || packing.count > INT_MAX || packing.carrier_count > INT_MAX)
            ckc::raise_status(ROCKE_ERR_VALUE, "invalid matrix fragment chunk layout");
        const rocke_type_t* unit_type = rocke_storage_ir_type(storage->dtype->name);
        if(!unit_type || ptr->type->kind != ROCKE_TYPE_PTR
           || !rocke_type_eq(ptr->type->pointee, unit_type))
            ckc::raise_status(ROCKE_ERR_VALUE, "matrix pointer storage type mismatch");
        const uint64_t unit_bytes = rocke_dtype_info(unit_type->name)->encoded_bits / 8;
        if(storage->packing.element_bits != packing.packing.element_bits
           || storage->packing.slot_bits != packing.packing.slot_bits || storage->base_bit_offset)
            ckc::raise_status(ROCKE_ERR_VALUE,
                              "matrix fragment requires matching packing and a zero bit origin");
        const uint64_t span = packing.count * layout->lane_groups;
        if(k0 > storage->cols || span > storage->cols - k0)
            ckc::raise_status(ROCKE_ERR_VALUE, "matrix fragment exceeds the packed row");
        uint64_t origin_bits, chunk_bytes;
        if(!rocke_bit_packing_offset(&storage->packing, k0, &origin_bits)
           || !rocke_bit_packing_bytes(&packing.packing, layout->chunk_elements, 0, &chunk_bytes)
           || origin_bits % (8 * unit_bytes) || chunk_bytes % unit_bytes)
            ckc::raise_status(ROCKE_ERR_VALUE,
                              "matrix fragment is not aligned to pointer storage units");
        const rocke_dtype_info_t* carrier_info = rocke_dtype_info(carrier_type->name);
        if(!carrier_info || carrier_info->encoded_bits != packing.carrier_bits)
            ckc::raise_status(ROCKE_ERR_VALUE, "matrix carrier type width mismatch");
        const uint64_t payload_bits = packing.count * packing.packing.slot_bits;
        if(payload_bits % packing.carrier_bits)
            ckc::raise_status(ROCKE_ERR_VALUE,
                              "matrix fragment payload must occupy whole carriers");
        const int live_carriers = payload_bits / packing.carrier_bits;
        const int padding = packing.carrier_count - live_carriers;
        if(padding && !rocke_type_eq(carrier_type, rocke_i32()))
            ckc::raise_status(ROCKE_ERR_VALUE,
                              "padded matrix fragments currently require i32 carriers");
        const uint64_t chunk_units = chunk_bytes / unit_bytes;
        const uint64_t origin_bytes = origin_bits / 8;
        if(origin_bytes / unit_bytes > INT_MAX || chunk_units * layout->lane_groups > INT_MAX)
            ckc::raise_status(ROCKE_ERR_VALUE, "matrix fragment offset exceeds i32 range");
        int alignment = gcd(gcd(storage->alignment_bytes, storage->row_stride_bytes),
                            gcd(chunk_bytes, origin_bytes));
        auto* lane_chunk = rocke_b_mul(b, lane_group, rocke_b_const_i32(b, chunk_units));
        auto* step_base = rocke_b_add(b, row_base, rocke_b_const_i32(b, origin_bytes / unit_bytes));
        /* Collect loads before concatenation to preserve Python SSA numbering. */
        rocke_value_t** chunks = (rocke_value_t**)rocke_arena_alloc(
            &b->arena, size_t(packing.count) * sizeof(rocke_value_t*));
        if(!chunks)
            ckc::raise_status(ROCKE_ERR_OOM, "matrix fragment allocation failed");
        int num_chunks = 0;
        for(uint64_t j = 0; j < packing.count / layout->chunk_elements; ++j)
        {
            auto* offset = rocke_b_add(
                b,
                rocke_b_add(
                    b, step_base, rocke_b_const_i32(b, j * layout->lane_groups * chunk_units)),
                lane_chunk);
            uint64_t remaining = chunk_units, consumed = 0;
            const int max_width = unit_bytes == 4 ? 8 : 16;
            while(remaining)
            {
                int width = 1;
                while(width < max_width && uint64_t(width * 2) <= remaining)
                    width *= 2;
                auto* at
                    = consumed ? rocke_b_add(b, offset, rocke_b_const_i32(b, consumed)) : offset;
                int load_align = gcd(alignment, consumed * unit_bytes);
                auto* value
                    = width == 1 ? rocke_b_vector_splat(
                                       b, rocke_b_global_load(b, ptr, at, unit_type, load_align), 1)
                                 : rocke_b_global_load_vN(b, ptr, at, unit_type, width, load_align);
                chunks[num_chunks++] = value;
                consumed += width;
                remaining -= width;
            }
        }
        auto* payload = chunks[0];
        for(int i = 1; i < num_chunks; ++i)
            payload = rocke_b_vec_concat(b, payload, chunks[i]);
        const auto* payload_type = rocke_vector_type(b, carrier_type, live_carriers);
        if(!rocke_type_eq(payload->type, payload_type))
            payload = rocke_b_bitcast(b, payload, payload_type);
        if(padding)
            payload = rocke_b_vec_concat(
                b, payload, rocke_b_vector_splat(b, rocke_b_const_i32(b, 0), padding));
        return payload;
    });
}

rocke_status_t rocke_h_pack_fragment_bits(rocke_ir_builder_t* b,
                                          rocke_load_bits_fn load_bits,
                                          void* ctx,
                                          const rocke_fragment_packing_t* fragment,
                                          rocke_value_t** words,
                                          size_t word_count)
{
    return ckc::guard_status(b, [&]() -> rocke_status_t {
        if(!rocke_i_live(b))
            return ROCKE_ERR_VALUE;
        rocke_fragment_packing_t checked;
        if(!fragment || !load_bits || !words
           || !rocke_fragment_packing_init(&checked,
                                           &fragment->packing,
                                           fragment->count,
                                           fragment->carrier_bits,
                                           fragment->carrier_count)
           || word_count != fragment->carrier_count || fragment->count > INT_MAX)
            ckc::raise_status(ROCKE_ERR_VALUE, "invalid pattern packing arguments");
        const auto& f = *fragment;
        if(f.carrier_bits != 32 && f.carrier_bits != 64)
            ckc::raise_status(ROCKE_ERR_VALUE,
                              "IR pattern packing currently requires i32 or i64 carriers");
        const auto* word_type = f.carrier_bits == 64 ? rocke_i64() : rocke_i32();
        auto constant = [&](uint64_t value) {
            return f.carrier_bits == 64 ? rocke_b_const_i64(b, int64_t(value))
                                        : rocke_b_const_i32(b, int32_t(value));
        };
        for(size_t j = 0; j < word_count; ++j)
            words[j] = constant(0);
        for(uint64_t j = 0; j < f.count; ++j)
        {
            auto* pattern = load_bits(b, (int)j, ctx);
            if(!pattern
               || (!rocke_type_eq(pattern->type, rocke_i8())
                   && !rocke_type_eq(pattern->type, rocke_i16())
                   && !rocke_type_eq(pattern->type, rocke_i32())
                   && !rocke_type_eq(pattern->type, rocke_i64())))
                ckc::raise_status(ROCKE_ERR_VALUE,
                                  "pattern packing requires unsigned patterns in integer carriers");
            const int bits = rocke_dtype_info(pattern->type->name)->encoded_bits;
            if(bits < f.packing.element_bits)
                ckc::raise_status(ROCKE_ERR_VALUE, "pattern carrier is smaller than encoded width");
            if(bits < f.carrier_bits)
                pattern = rocke_b_zext(b, pattern, word_type);
            else if(bits > f.carrier_bits)
                ckc::raise_status(ROCKE_ERR_VALUE, "pattern carrier is wider than output carrier");
            uint64_t start = j * f.packing.slot_bits;
            int remaining = f.packing.element_bits, consumed = 0;
            while(remaining)
            {
                const uint64_t word = start / f.carrier_bits;
                const int shift = start % f.carrier_bits;
                const int take
                    = remaining < f.carrier_bits - shift ? remaining : f.carrier_bits - shift;
                auto* part = consumed ? rocke_b_lshr(b, pattern, constant(consumed)) : pattern;
                if(take < remaining)
                    part = rocke_b_land(b, part, constant((uint64_t(1) << take) - 1));
                words[word] = rocke_b_lor(b, words[word], rocke_b_shl(b, part, constant(shift)));
                start += take;
                consumed += take;
                remaining -= take;
            }
        }
        return ROCKE_OK;
    });
}
